### DISCUSSION TITLE
gadgetron running using [ENTRYPOINT]

### DISCUSSION THREAD

**Discussion by ahsanjav** [gadgetron/gadgetron#1133] [Q&A]
Is gadgetron thats launched using `ENTRYPOINT [ "/tini", "--", "/opt/entrypoint.sh" ]` running as root ? I don't think it is but I am unable to launch an external Matlab recon unless I manually start gadgetron from command line. 

I may need some help getting this to work.

**Comment by hansenms** [comment]
It is not running as root. It is running as a non-privileged user with UID 1000. Are you trying to also run Matlab inside the container?

**Reply by ahsanjav** [reply]
Yes - we have a few use cases where Matlab is launched in active mode as part of the reconstruction using the external interface. 

if I launch gadgetron from a bash shell within the container the reconstruction work fine however they fail when I use gadgetron which is running using the entrypoint. The output is not so helpful and I am not sure how to get more information to debug ...

Here is the log of the error. 

07-25 17:12:19.491 DEBUG [ConfigConnection.cpp:73] Reading config file: "/opt/conda/envs/gadgetron/share/gadgetron/config/external_matlab_tiny_example.xml"
07-25 17:12:19.493 INFO [HeaderConnection.cpp:84] Connection state: [HEADER]
07-25 17:12:19.494 INFO [StreamConnection.cpp:75] Connection state: [STREAM]
07-25 17:12:19.524 DEBUG [Stream.cpp:71] Loading External Execute block with name gadgetron.examples.tiny_recon of type matlab
07-25 17:12:19.525 DEBUG [Stream.cpp:58] Loading Gadget  of class NoiseAdjustGadget from gadgetron_mricore
07-25 17:12:19.525 DEBUG [Stream.cpp:58] Loading Gadget  of class RemoveROOversamplingGadget from gadgetron_mricore
07-25 17:12:19.525 INFO [External.cpp:77] Waiting for external module 'gadgetron.examples.tiny_recon' on port: 38745
07-25 17:12:19.526 DEBUG [NoiseAdjustGadget.cpp:195] Folder to store noise dependencies is /tmp/gadgetron
07-25 17:12:19.526 DEBUG [NoiseAdjustGadget.cpp:196] NoiseAdjustGadget::perform_noise_adjust_ is 1
07-25 17:12:19.526 DEBUG [NoiseAdjustGadget.cpp:197] NoiseAdjustGadget::pass_nonconformant_data_ is 1
07-25 17:12:19.526 DEBUG [NoiseAdjustGadget.cpp:198] receiver_noise_bandwidth_ is 0.793000
07-25 17:12:19.526 DEBUG [Stream.cpp:58] Loading Gadget  of class AcquisitionAccumulateTriggerGadget from gadgetron_mricore
07-25 17:12:19.527 DEBUG [Stream.cpp:58] Loading Gadget  of class BucketToBufferGadget from gadgetron_mricore
07-25 17:12:19.530 INFO [Matlab.cpp:31] Started external MATLAB module (pid: 3390).
**07-25 17:12:19.532 ERROR [Core.cpp:18] [Connection Main Thread/Stream/external] ERROR: Operation canceled**
07-25 17:12:19.532 DEBUG [RemoveROOversamplingGadget.cpp:26] RemoveROOversamplingGadget:omp_set_num_threads(1) ...
07-25 17:12:19.532 DEBUG [NoiseAdjustGadget.cpp:210] Measurement ID is
07-25 17:12:19.550 DEBUG [AcquisitionAccumulateTriggerGadget.cpp:129] Trigger (1) occurred, sending out 1 buckets
07-25 17:12:19.550 DEBUG [BucketToBufferGadget.cpp:50] BUCKET_SIZE 12 ESPACE 0
07-25 17:12:19.557 DEBUG [AcquisitionAccumulateTriggerGadget.cpp:129] Trigger (2) occurred, sending out 1 buckets
07-25 17:12:19.561 DEBUG [BucketToBufferGadget.cpp:90] End of bucket reached, sending out 1 ReconData buffers
07-25 17:12:19.570 DEBUG [AcquisitionAccumulateTriggerGadget.cpp:129] Trigger (3) occurred, sending out 1 buckets
07-25 17:12:19.574 INFO [Core.cpp:77] Connection state: [FINISHED]

**Reply by hansenms** [reply]
When you say "launch from a bash shell", how exactly do you do that? I suspect there is something going on with the Gadgetron running in a conda environment (which is activated in the `/opt/entrypoint.sh` script).

**Reply by ahsanjav** [reply]
On the server I run : docker exec -ti gt_test bash
in the shell which is running in the conda environment. I run **gadgetron**.

**Reply by hansenms** [reply]
So you are starting a second Gadgetron instance inside a container already running Gadgetron? What if you do `docker run --rm -ti --entrypoint /bin/bash ghcr.io/gadgetron/gadgetron/gadgetron_ubuntu_rt_cuda:latest` and run `gadgetron`, does that work? (replace the image name with whatever your image is called).

**Reply by ahsanjav** [reply]
That works.

**Reply by ahsanjav** [reply]
I think it just has to do with setting the matlab env variable which is not happening in entrypoint.sh - I don't quite understand how paths and env variables are set for programs exec in entrypoint.sh - but I have a solution . I can either add the paths and variables to the end of conda.sh or in another script and execute it similar to conda.sh and it seems to fix the issue.

**Reply by hansenms** [reply]
You can modify the entrypoint.sh script to set additional environment variables. You should also be able to set those environment variables on the Docker container when you do Docker run.
